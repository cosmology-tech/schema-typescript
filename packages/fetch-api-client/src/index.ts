import fetch from 'isomorphic-fetch';

interface RequestOptions<Params> {
  hostname: string;
  path: string;
  headers?: { [key: string]: string };
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
  params?: Params;
  timeout?: number;
}

export interface APIClientOptions {
  restEndpoint: string;
}

export class APIClient {
  private baseUrl: string;
  private defaultTimeout: number = 10000; // 10 seconds

  constructor(options: APIClientOptions) {
    this.baseUrl = options.restEndpoint;
  }

  private buildUrl(endpoint: string, query?: { [key: string]: any }): string {
    const url = new URL(endpoint, this.baseUrl);
    if (query) {
      Object.keys(query).forEach(key => url.searchParams.append(key, query[key]));
    }
    return url.toString();
  }

  private async request<Resp>(options: RequestOptions<any>): Promise<Resp> {
    const { path, headers, method, params, timeout } = options;
    const url = this.buildUrl(path, method === 'GET' && params);
    const fetchOptions: RequestInit = {
      method,
      headers,
      body: method !== 'GET' && method !== 'DELETE' ? JSON.stringify(params) : null,
    };

    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), timeout || this.defaultTimeout);
    fetchOptions.signal = controller.signal;

    const response = await fetch(url, fetchOptions);
    clearTimeout(id);

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    return response.json() as Promise<Resp>;
  }

  get<Resp = unknown>(endpoint: string, query?: { [key: string]: any }, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'GET',
      headers: opts.headers,
      timeout: opts.timeout || this.defaultTimeout,
      params: query, // Correctly use query parameters for GET
      hostname: '',
    });
  }

  post<Resp = unknown, Params = any>(endpoint: string, params: Params, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...opts.headers,
      },
      params,
      timeout: opts.timeout || this.defaultTimeout,
      hostname: '',
    });
  }

  put<Resp = unknown, Params = any>(endpoint: string, params: Params, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        ...opts.headers,
      },
      params,
      timeout: opts.timeout || this.defaultTimeout,
      hostname: '',
    });
  }

  patch<Resp = unknown, Params = any>(endpoint: string, params: Params, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        ...opts.headers,
      },
      params,
      timeout: opts.timeout || this.defaultTimeout,
      hostname: '',
    });
  }

  delete<Resp = unknown>(endpoint: string, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'DELETE',
      headers: opts.headers,
      timeout: opts.timeout || this.defaultTimeout,
      hostname: '',
    });
  }
}
