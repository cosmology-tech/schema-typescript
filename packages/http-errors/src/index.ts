export class HTTPError extends Error {
  constructor(message: string, public status: number) {
    super(`${message} | Status: ${status}`);
    this.name = this.constructor.name;
  }
}
    
export class BadRequestError extends HTTPError {
  constructor() { super('Bad Request', 400); }
}
export class UnauthorizedError extends HTTPError {
  constructor() { super('Unauthorized', 401); }
}
export class PaymentRequiredError extends HTTPError {
  constructor() { super('Payment Required', 402); }
}
export class ForbiddenError extends HTTPError {
  constructor() { super('Forbidden', 403); }
}
export class NotFoundError extends HTTPError {
  constructor() { super('Not Found', 404); }
}
export class MethodNotAllowedError extends HTTPError {
  constructor() { super('Method Not Allowed', 405); }
}
export class NotAcceptableError extends HTTPError {
  constructor() { super('Not Acceptable', 406); }
}
export class ProxyAuthenticationRequiredError extends HTTPError {
  constructor() { super('Proxy Authentication Required', 407); }
}
export class RequestTimeoutError extends HTTPError {
  constructor() { super('Request Timeout', 408); }
}
export class ConflictError extends HTTPError {
  constructor() { super('Conflict', 409); }
}
export class GoneError extends HTTPError {
  constructor() { super('Gone', 410); }
}
export class LengthRequiredError extends HTTPError {
  constructor() { super('Length Required', 411); }
}
export class PreconditionFailedError extends HTTPError {
  constructor() { super('Precondition Failed', 412); }
}
export class PayloadTooLargeError extends HTTPError {
  constructor() { super('Payload Too Large', 413); }
}
export class URITooLongError extends HTTPError {
  constructor() { super('URI Too Long', 414); }
}
export class UnsupportedMediaTypeError extends HTTPError {
  constructor() { super('Unsupported Media Type', 415); }
}
export class RangeNotSatisfiableError extends HTTPError {
  constructor() { super('Range Not Satisfiable', 416); }
}
export class ExpectationFailedError extends HTTPError {
  constructor() { super('Expectation Failed', 417); }
}
export class ImATeapotError extends HTTPError {
  constructor() { super("I'm a Teapot", 418); }
}
export class MisdirectedRequestError extends HTTPError {
  constructor() { super('Misdirected Request', 421); }
}
export class UnprocessableEntityError extends HTTPError {
  constructor() { super('Unprocessable Entity', 422); }
}
export class LockedError extends HTTPError {
  constructor() { super('Locked', 423); }
}
export class FailedDependencyError extends HTTPError {
  constructor() { super('Failed Dependency', 424); }
}
export class UpgradeRequiredError extends HTTPError {
  constructor() { super('Upgrade Required', 426); }
}
export class PreconditionRequiredError extends HTTPError {
  constructor() { super('Precondition Required', 428); }
}
export class TooManyRequestsError extends HTTPError {
  constructor() { super('Too Many Requests', 429); }
}
export class RequestHeaderFieldsTooLargeError extends HTTPError {
  constructor() { super('Request Header Fields Too Large', 431); }
}
export class UnavailableForLegalReasonsError extends HTTPError {
  constructor() { super('Unavailable For Legal Reasons', 451); }
}
export class InternalServerError extends HTTPError {
  constructor() { super('Internal Server Error', 500); }
}
export class NotImplementedError extends HTTPError {
  constructor() { super('Not Implemented', 501); }
}
export class BadGatewayError extends HTTPError {
  constructor() { super('Bad Gateway', 502); }
}
export class ServiceUnavailableError extends HTTPError {
  constructor() { super('Service Unavailable', 503); }
}
export class GatewayTimeoutError extends HTTPError {
  constructor() { super('Gateway Timeout', 504); }
}
export class HTTPVersionNotSupportedError extends HTTPError {
  constructor() { super('HTTP Version Not Supported', 505); }
}
export class VariantAlsoNegotiatesError extends HTTPError {
  constructor() { super('Variant Also Negotiates', 506); }
}
export class InsufficientStorageError extends HTTPError {
  constructor() { super('Insufficient Storage', 507); }
}
export class LoopDetectedError extends HTTPError {
  constructor() { super('Loop Detected', 508); }
}
export class NotExtendedError extends HTTPError {
  constructor() { super('Not Extended', 510); }
}
export class NetworkAuthenticationRequiredError extends HTTPError {
  constructor() { super('Network Authentication Required', 511); }
}
    
  
export function createHTTPError(status: number): HTTPError {
  switch (status) {
  case 400: return new BadRequestError();
  case 401: return new UnauthorizedError();
  case 402: return new PaymentRequiredError();
  case 403: return new ForbiddenError();
  case 404: return new NotFoundError();
  case 405: return new MethodNotAllowedError();
  case 406: return new NotAcceptableError();
  case 407: return new ProxyAuthenticationRequiredError();
  case 408: return new RequestTimeoutError();
  case 409: return new ConflictError();
  case 410: return new GoneError();
  case 411: return new LengthRequiredError();
  case 412: return new PreconditionFailedError();
  case 413: return new PayloadTooLargeError();
  case 414: return new URITooLongError();
  case 415: return new UnsupportedMediaTypeError();
  case 416: return new RangeNotSatisfiableError();
  case 417: return new ExpectationFailedError();
  case 418: return new ImATeapotError();
  case 421: return new MisdirectedRequestError();
  case 422: return new UnprocessableEntityError();
  case 423: return new LockedError();
  case 424: return new FailedDependencyError();
  case 426: return new UpgradeRequiredError();
  case 428: return new PreconditionRequiredError();
  case 429: return new TooManyRequestsError();
  case 431: return new RequestHeaderFieldsTooLargeError();
  case 451: return new UnavailableForLegalReasonsError();
  case 500: return new InternalServerError();
  case 501: return new NotImplementedError();
  case 502: return new BadGatewayError();
  case 503: return new ServiceUnavailableError();
  case 504: return new GatewayTimeoutError();
  case 505: return new HTTPVersionNotSupportedError();
  case 506: return new VariantAlsoNegotiatesError();
  case 507: return new InsufficientStorageError();
  case 508: return new LoopDetectedError();
  case 510: return new NotExtendedError();
  case 511: return new NetworkAuthenticationRequiredError();
  default: return new HTTPError('Unknown Error', status);
  }
}
  