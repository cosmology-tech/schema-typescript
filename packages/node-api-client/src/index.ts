import * as http from 'http';
import * as querystring from 'querystring';
import { URLSearchParams } from 'url';

interface RequestOptions<Params> {
  hostname: string;
  path: string;
  headers?: { [key: string]: string };
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
  params?: Params;
  timeout?: number;
}

export interface APIClientOptions {
  restEndpoint: string;
}

export interface APIClientRequestHeaders {
  accept?: string | string[] | undefined;
  'accept-charset'?: string | string[] | undefined;
  'accept-encoding'?: string | string[] | undefined;
  'accept-language'?: string | string[] | undefined;
  'accept-ranges'?: string | undefined;
  'access-control-allow-credentials'?: string | undefined;
  'access-control-allow-headers'?: string | undefined;
  'access-control-allow-methods'?: string | undefined;
  'access-control-allow-origin'?: string | undefined;
  'access-control-expose-headers'?: string | undefined;
  'access-control-max-age'?: string | undefined;
  'access-control-request-headers'?: string | undefined;
  'access-control-request-method'?: string | undefined;
  age?: string | undefined;
  allow?: string | undefined;
  authorization?: string | undefined;
  'cache-control'?: string | undefined;
  'cdn-cache-control'?: string | undefined;
  connection?: string | string[] | undefined;
  'content-disposition'?: string | undefined;
  'content-encoding'?: string | undefined;
  'content-language'?: string | undefined;
  'content-length'?: string | number | undefined;
  'content-location'?: string | undefined;
  'content-range'?: string | undefined;
  'content-security-policy'?: string | undefined;
  'content-security-policy-report-only'?: string | undefined;
  cookie?: string | string[] | undefined;
  dav?: string | string[] | undefined;
  dnt?: string | undefined;
  date?: string | undefined;
  etag?: string | undefined;
  expect?: string | undefined;
  expires?: string | undefined;
  forwarded?: string | undefined;
  from?: string | undefined;
  host?: string | undefined;
  'if-match'?: string | undefined;
  'if-modified-since'?: string | undefined;
  'if-none-match'?: string | undefined;
  'if-range'?: string | undefined;
  'if-unmodified-since'?: string | undefined;
  'last-modified'?: string | undefined;
  link?: string | string[] | undefined;
  location?: string | undefined;
  'max-forwards'?: string | undefined;
  origin?: string | undefined;
  prgama?: string | string[] | undefined;
  'proxy-authenticate'?: string | string[] | undefined;
  'proxy-authorization'?: string | undefined;
  'public-key-pins'?: string | undefined;
  'public-key-pins-report-only'?: string | undefined;
  range?: string | undefined;
  referer?: string | undefined;
  'referrer-policy'?: string | undefined;
  refresh?: string | undefined;
  'retry-after'?: string | undefined;
  'sec-websocket-accept'?: string | undefined;
  'sec-websocket-extensions'?: string | string[] | undefined;
  'sec-websocket-key'?: string | undefined;
  'sec-websocket-protocol'?: string | string[] | undefined;
  'sec-websocket-version'?: string | undefined;
  server?: string | undefined;
  'set-cookie'?: string | string[] | undefined;
  'strict-transport-security'?: string | undefined;
  te?: string | undefined;
  trailer?: string | undefined;
  'transfer-encoding'?: string | undefined;
  'user-agent'?: string | undefined;
  upgrade?: string | undefined;
  'upgrade-insecure-requests'?: string | undefined;
  vary?: string | undefined;
  via?: string | string[] | undefined;
  warning?: string | undefined;
  'www-authenticate'?: string | string[] | undefined;
  'x-content-type-options'?: string | undefined;
  'x-dns-prefetch-control'?: string | undefined;
  'x-frame-options'?: string | undefined;
  'x-xss-protection'?: string | undefined;
}
export type APIClientRequestOpts = {
  headers?: APIClientRequestHeaders;
  timeout?: number;
};
export class APIClient {
  private hostname: string;
  private port: number = 8001; // Default Kubernetes proxy port
  private defaultTimeout: number = 10000; // 10 seconds

  constructor(options: APIClientOptions) {
    const url = new URL(options.restEndpoint);
    this.hostname = url.hostname;
    this.port = url.port ? parseInt(url.port) : this.port;
  }

  private buildFullPath(endpoint: string, query?: { [key: string]: any }): string {
    const queryString = query ? `?${querystring.stringify(query)}` : '';
    return endpoint + queryString;
  }

  get<Resp = unknown>(
    endpoint: string,
    query?: { [key: string]: any },
    body?: void,
    opts: APIClientRequestOpts = {}
  ): Promise<Resp> {
    const fullPath = this.buildFullPath(endpoint, query);
    return this.request<Resp, void>({
      path: fullPath,
      method: 'GET',
      // @ts-ignore
      headers: opts.headers,
      timeout: opts.timeout || this.defaultTimeout,
    });
  }

  post<Resp = unknown, Params = any>(
    endpoint: string,
    query?: { [key: string]: any },
    body?: Params,
    opts: http.RequestOptions & { timeout?: number; isFormData?: boolean } = {}
  ): Promise<Resp> {
    const fullPath = this.buildFullPath(endpoint, query);
    const headers: http.OutgoingHttpHeaders = opts.isFormData
      ? {
        'Content-Type': 'application/x-www-form-urlencoded',
        ...opts.headers,
      }
      : {
        'Content-Type': 'application/json',
        ...opts.headers,
      };
    const bodyContent = opts.isFormData
      ? new URLSearchParams(body as any).toString()
      : JSON.stringify(body);
    headers['Content-Length'] = Buffer.byteLength(bodyContent).toString();

    return this.request<Resp, Params>({
      path: fullPath,
      method: 'POST',
      // @ts-ignore
      headers,
      // @ts-ignore
      params: bodyContent,
      timeout: opts.timeout || this.defaultTimeout,
    });
  }

  patch<Resp = unknown, Params = any>(
    endpoint: string,
    query?: { [key: string]: any },
    body?: Params,
    opts: http.RequestOptions & { timeout?: number; isFormData?: boolean } = {}
  ): Promise<Resp> {
    const fullPath = this.buildFullPath(endpoint, query);
    const headers: http.OutgoingHttpHeaders = opts.isFormData
      ? {
        'Content-Type': 'application/x-www-form-urlencoded',
        ...opts.headers,
      }
      : {
        'Content-Type': 'application/json',
        ...opts.headers,
      };
    const bodyContent = opts.isFormData
      ? new URLSearchParams(body as any).toString()
      : JSON.stringify(body);
    headers['Content-Length'] = Buffer.byteLength(bodyContent).toString();

    return this.request<Resp, Params>({
      path: fullPath,
      method: 'PATCH',
      // @ts-ignore
      headers,
      // @ts-ignore
      params: bodyContent,
      timeout: opts.timeout || this.defaultTimeout,
    });
  }

  put<Resp = unknown, Params = any>(
    endpoint: string,
    query?: { [key: string]: any },
    body?: Params,
    opts: http.RequestOptions & { timeout?: number } = {}
  ): Promise<Resp> {
    const fullPath = this.buildFullPath(endpoint, query);
    const bodyContent = JSON.stringify(body);
    const headers = {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(bodyContent).toString(),
      ...opts.headers,
    };

    return this.request<Resp, Params>({
      path: fullPath,
      method: 'PUT',
      // @ts-ignore
      headers,
      // @ts-ignore
      params: bodyContent,
      timeout: opts.timeout || this.defaultTimeout,
    });
  }

  delete<Resp = unknown, Params = any>(
    endpoint: string,
    query?: { [key: string]: any },
    body?: void,
    opts: http.RequestOptions & { timeout?: number } = {}
  ): Promise<Resp> {
    const fullPath = this.buildFullPath(endpoint, query);
    return this.request<Resp, void>({
      path: fullPath,
      method: 'DELETE',
      // @ts-ignore
      headers: opts.headers,
      timeout: opts.timeout || this.defaultTimeout,
    });
  }

  private request<Resp, Params>(
    options: RequestOptions<Params>
  ): Promise<Resp> {
    return new Promise((resolve, reject) => {
      const requestOptions: http.RequestOptions = {
        hostname: this.hostname,
        port: this.port,
        path: options.path,
        method: options.method,
        headers: options.headers,
      };

      const req = http.request(requestOptions, (res) => {
        let data = '';
        res.on('data', (chunk) => {
          data += chunk;
        });
        res.on('end', () => {
          try {
            const parsedData: Resp = JSON.parse(data);
            resolve(parsedData);
          } catch (error) {
            reject(error);
          }
        });
      });

      req.on('error', (error) => {
        reject(error);
      });

      req.setTimeout(options.timeout, () => {
        req.abort();
        reject(new Error('Request timeout'));
      });

      if (options.params && ['POST', 'PUT', 'PATCH'].includes(options.method)) {
        req.write(options.params);
      }

      req.end();
    });
  }
}
