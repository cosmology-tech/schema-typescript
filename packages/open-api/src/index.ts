export * from './openapi';
export * from './openapi.types';
export * from './utils';