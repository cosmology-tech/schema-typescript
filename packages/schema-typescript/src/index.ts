export * from './context';
export * from './schema';
export * from './types';
export * from './utils';
