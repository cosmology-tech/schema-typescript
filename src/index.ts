// @ts-nocheck
import * as t from "@babel/types";
import generate from "@babel/generator";
import { toPascalCase } from "./utils";

interface JSONSchema {
  title: string;
  properties?: { [key: string]: JSONSchemaProperty };
  required?: string[];
  type?: string;
  items?: JSONSchema;
  $defs?: { [key: string]: JSONSchema };
}

type JSONSchemaProperty = {
  type?: string;
  properties?: { [key: string]: JSONSchemaProperty };
  items?: JSONSchemaProperty;
  required?: string[];
  $ref?: string;
};

interface SchemaTSOptions {
  useSingleQuotes: boolean;
}

interface SchemaTSContextI {
  options: SchemaTSOptions;
  root: JSONSchema;
  schema: JSONSchema;
  parents: JSONSchema[]
}


class SchemaTSContext implements SchemaTSContextI {
  options: SchemaTSOptions;
  root: JSONSchema;
  schema: JSONSchema;
  parents: JSONSchema[];

  constructor(
    options: SchemaTSOptions,
    root: JSONSchema,
    schema: JSONSchema,
    parents: JSONSchema[] = []
  ) {
    this.options = options;
    this.schema = schema;
    this.root = root;
    this.parents = parents;
  }

  // Clone the context with the option to add a new parent
  clone(newParent: JSONSchema): Context {
    // Create a new array for parents to avoid mutation of the original array
    const newParents = [...this.parents, newParent];
    return new SchemaTSContext(this.options, this.root, this.schema, newParents);
  }
}

const defaultOptions: SchemaTSOptions = { useSingleQuotes: true };

export function generateTypeScript(schema: JSONSchema, options?: SchemaTSOptions): string {
  const interfaces = [];
  const opts = options || defaultOptions;
  const ctx = new SchemaTSContext(opts, schema);

  try {

    // Process definitions first
    if (schema.$defs) {
      for (const key in schema.$defs) {
        // asset requires denom, but is also a definition.
        // maybe we just need to do a two-pass and register
        interfaces.push(createInterfaceDeclaration(ctx, toPascalCase(key), schema.$defs[key]));
      }
    }
  } catch (e) {
    console.error('Error processing interfaces');
    throw e;
  }
  // Process the main schema
  interfaces.push(createInterfaceDeclaration(ctx, toPascalCase(schema.title), schema));
  return generate(t.file(t.program(interfaces))).code;
}

function createInterfaceDeclaration(ctx: SchemaTSContext, name: string, schema: JSONSchema): t.TSInterfaceDeclaration {
  const properties = schema.properties || {};
  const required = schema.required || [];
  const body = Object.keys(properties).map(key => {
    const prop = properties[key];
    return createPropertySignature(ctx, key, prop, required, schema);
  });

  return t.tsInterfaceDeclaration(
    t.identifier(name),
    null,
    [],
    t.tsInterfaceBody(body)
  );
}

// Determine if the key is a valid JavaScript identifier
function isValidIdentifier(key) {
  return /^[$A-Z_][0-9A-Z_$]*$/i.test(key) && !/^[0-9]+$/.test(key);
}

function createPropertySignature(
  ctx: SchemaTSContext,
  key: string,
  prop: JSONSchemaProperty,
  required: string[],
  schema: JSONSchema
): t.TSPropertySignature {

  // Adjust the quoting style based on the useSingleQuotes flag
  function formatKey(ctx: SchemaTSContext) {
    const quote = ctx.options.useSingleQuotes ? "'" : '"';
    return `${quote}${key}${quote}`;
  }

  const propType = getTypeForProp(ctx, prop, required, schema);
  const identifier = isValidIdentifier(key) ? t.identifier(key) : t.stringLiteral(formatKey(ctx, key));
  const propSig = t.tsPropertySignature(
    identifier,
    t.tsTypeAnnotation(propType)
  );
  propSig.optional = !required.includes(key);
  return propSig;
}


function getTypeForProp(ctx: SchemaTSContext, prop: JSONSchemaProperty, required: string[], schema: JSONSchema): t.TSType {
  if (prop.$ref) {
    return resolveRefType(ctx, prop.$ref, schema);
  }

  switch (prop.type) {
    case 'string':
      return t.tsStringKeyword();
    case 'number':
    case 'integer':
      return t.tsNumberKeyword();
    case 'boolean':
      return t.tsBooleanKeyword();
    case 'array':
      if (prop.items) {
        return t.tsArrayType(getTypeForProp(ctx, prop.items, required, schema));
      } else {
        throw new Error('Array items specification is missing');
      }
    case 'object':
      if (prop.properties) {
        const nestedProperties = prop.properties;
        const nestedRequired = prop.required || [];
        const typeElements = Object.keys(nestedProperties).map(nestedKey => {
          const nestedProp = nestedProperties[nestedKey];
          return createPropertySignature(ctx, nestedKey, nestedProp, nestedRequired, schema);
        });
        return t.tsTypeLiteral(typeElements);
      } else {
        throw new Error('Object must have properties');
      }
    default:
      return t.tsAnyKeyword();
  }
}

function resolveRefType(ctx: SchemaTSContext, ref: string, schema: JSONSchema): t.TSType {
  const path = ref.split('/');
  const definitionName = path.pop();
  if (definitionName && schema.$defs && schema.$defs[definitionName]) {
    return t.tsTypeReference(t.identifier(toPascalCase(definitionName)));
  }
  // now look in root
  if (definitionName && ctx.root.$defs && ctx.root.$defs[definitionName]) {
    return t.tsTypeReference(t.identifier(toPascalCase(definitionName)));
  }

  throw new Error(`Reference ${ref} not found in definitions.`);
}
