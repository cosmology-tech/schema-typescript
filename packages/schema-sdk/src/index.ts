import { getDefaultSchemaTSOptions } from 'schema-typescript';

export * from './openapi';
export * from './openapi.types';
export * from './utils';

export { getDefaultSchemaTSOptions };