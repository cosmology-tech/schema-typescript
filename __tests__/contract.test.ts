import { writeFileSync } from 'fs';

import schema from '../__fixtures__/schemas/contract.json';
import { generateTypeScript } from '../src';

it('contract', () => {
    const code = generateTypeScript(schema as any);
    expect(code).toMatchSnapshot();
    writeFileSync(__dirname + '/../__fixtures__/output/contract.ts', code);
});

it('contract camelCase', () => {
    const code = generateTypeScript(schema as any, {
        useSingleQuotes: true,
        camelCase: true
    });
    expect(code).toMatchSnapshot();
    writeFileSync(__dirname + '/../__fixtures__/output/contract.camel.ts', code);
});
