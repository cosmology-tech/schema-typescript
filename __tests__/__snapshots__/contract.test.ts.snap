// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contract 1`] = `
"export type Addr = string;
export interface Asset {
  amount: Uint128;
  info: AssetInfo;
}
export type AssetInfo = {
  token: {
    contract_addr: Addr;
  };
} | {
  native_token: {
    denom: string;
  };
};
export type Binary = string;
export interface Cw20ReceiveMsg {
  amount: Uint128;
  msg: Binary;
  sender: string;
}
export type Decimal = string;
export interface FeeInfo {
  dev_fee_percent: number;
  developer_addr?: Addr | any;
  protocol_fee_percent: number;
  total_fee_bps: number;
}
export interface PoolConfig {
  code_id: number;
  fee_info: FeeInfo;
  is_disabled: boolean;
  is_generator_disabled: boolean;
  pool_type: PoolType;
}
export type PoolType = {
  xyk: any;
} | {
  stable2_pool: any;
} | {
  stable3_pool: any;
} | {
  weighted: any;
} | {
  custom: string;
};
export interface SingleSwapRequest {
  amount: Uint128;
  asset_in: AssetInfo;
  asset_out: AssetInfo;
  belief_price?: Decimal | any;
  max_spread?: Decimal | any;
  pool_id: Uint128;
  swap_type: SwapType;
}
export type SwapType = {
  give_in: any;
} | {
  give_out: any;
} | {
  custom: string;
};
export type Uint128 = string;
export type ExecuteMsg = {
  receive: Cw20ReceiveMsg;
} | {
  update_config: {
    fee_collector?: any;
    generator_address?: any;
    lp_token_code_id?: any;
  };
} | {
  update_pool_config: {
    is_disabled?: any;
    new_fee_info?: FeeInfo | any;
    pool_type: PoolType;
  };
} | {
  add_to_registery: {
    new_pool_config: PoolConfig;
  };
} | {
  create_pool_instance: {
    asset_infos: AssetInfo[];
    init_params?: Binary | any;
    lp_token_name?: any;
    lp_token_symbol?: any;
    pool_type: PoolType;
  };
} | {
  join_pool: {
    assets?: any;
    auto_stake?: any;
    lp_to_mint?: Uint128 | any;
    pool_id: Uint128;
    recipient?: any;
    slippage_tolerance?: Decimal | any;
  };
} | {
  swap: {
    recipient?: any;
    swap_request: SingleSwapRequest;
  };
} | {
  propose_new_owner: {
    expires_in: number;
    owner: string;
  };
} | {
  drop_ownership_proposal: any;
} | {
  claim_ownership: any;
};"
`;

exports[`contract camelCase 1`] = `
"export type Addr = string;
export interface Asset {
  amount: Uint128;
  info: AssetInfo;
}
export type AssetInfo = {
  token: {
    contractAddr: Addr;
  };
} | {
  nativeToken: {
    denom: string;
  };
};
export type Binary = string;
export interface Cw20ReceiveMsg {
  amount: Uint128;
  msg: Binary;
  sender: string;
}
export type Decimal = string;
export interface FeeInfo {
  devFeePercent: number;
  developerAddr?: Addr | any;
  protocolFeePercent: number;
  totalFeeBps: number;
}
export interface PoolConfig {
  codeId: number;
  feeInfo: FeeInfo;
  isDisabled: boolean;
  isGeneratorDisabled: boolean;
  poolType: PoolType;
}
export type PoolType = {
  xyk: any;
} | {
  stable2Pool: any;
} | {
  stable3Pool: any;
} | {
  weighted: any;
} | {
  custom: string;
};
export interface SingleSwapRequest {
  amount: Uint128;
  assetIn: AssetInfo;
  assetOut: AssetInfo;
  beliefPrice?: Decimal | any;
  maxSpread?: Decimal | any;
  poolId: Uint128;
  swapType: SwapType;
}
export type SwapType = {
  giveIn: any;
} | {
  giveOut: any;
} | {
  custom: string;
};
export type Uint128 = string;
export type ExecuteMsg = {
  receive: Cw20ReceiveMsg;
} | {
  updateConfig: {
    feeCollector?: any;
    generatorAddress?: any;
    lpTokenCodeId?: any;
  };
} | {
  updatePoolConfig: {
    isDisabled?: any;
    newFeeInfo?: FeeInfo | any;
    poolType: PoolType;
  };
} | {
  addToRegistery: {
    newPoolConfig: PoolConfig;
  };
} | {
  createPoolInstance: {
    assetInfos: AssetInfo[];
    initParams?: Binary | any;
    lpTokenName?: any;
    lpTokenSymbol?: any;
    poolType: PoolType;
  };
} | {
  joinPool: {
    assets?: any;
    autoStake?: any;
    lpToMint?: Uint128 | any;
    poolId: Uint128;
    recipient?: any;
    slippageTolerance?: Decimal | any;
  };
} | {
  swap: {
    recipient?: any;
    swapRequest: SingleSwapRequest;
  };
} | {
  proposeNewOwner: {
    expiresIn: number;
    owner: string;
  };
} | {
  dropOwnershipProposal: any;
} | {
  claimOwnership: any;
};"
`;
